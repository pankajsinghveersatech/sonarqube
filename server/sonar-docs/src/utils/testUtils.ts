/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
