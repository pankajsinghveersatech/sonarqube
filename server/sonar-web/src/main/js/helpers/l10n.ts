/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10n';
import { toNotSoISOString } from './dates';
import { get as loadFromLocalStorage, save as saveInLocalStorage } from './storage';

export type Messages = T.Dict<string>;

export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.'
};

let allMessages: Messages = {};
let locale: string | undefined;

export function translate(...keys: string[]): string {
  const messageKey = keys.join('.');
  const l10nMessages = getMessages();
  if (process.env.NODE_ENV === 'development' && !l10nMessages[messageKey]) {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return l10nMessages[messageKey] || messageKey;
}

export function translateWithParameters(
  messageKey: string,
  ...parameters: Array<string | number>
): string {
  const message = getMessages()[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, () => parameter), message);
  }
  if (process.env.NODE_ENV === 'development') {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return `${messageKey}.${parameters.join('.')}`;
}

export function hasMessage(...keys: string[]): boolean {
  const messageKey = keys.join('.');
  return getMessages()[messageKey] != null;
}

export function getMessages() {
  if (typeof allMessages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return allMessages;
}

export function resetMessages(newMessages: Messages) {
  allMessages = newMessages;
}

export function getLocalizedMetricName(
  metric: { key: string; name?: string },
  short = false
): string {
  const bundleKey = `metric.${metric.key}.${short ? 'short_name' : 'name'}`;
  if (hasMessage(bundleKey)) {
    return translate(bundleKey);
  } else if (short) {
    return getLocalizedMetricName(metric);
  }
  return metric.name || metric.key;
}

export function getLocalizedCategoryMetricName(metric: { key: string; name?: string }) {
  const bundleKey = `metric.${metric.key}.extra_short_name`;
  return hasMessage(bundleKey) ? translate(bundleKey) : getLocalizedMetricName(metric, true);
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  return hasMessage(bundleKey) ? translate(bundleKey) : domainName;
}

export function getCurrentLocale() {
  return locale;
}

export function resetCurrentLocale(newLocale: string) {
  locale = newLocale;
}

export function getShortMonthName(index: number) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ];
  return translate(months[index]);
}

export function getWeekDayName(index: number) {
  const weekdays = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

export function getShortWeekDayName(index: number) {
  const weekdays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

const L10N_BUNDLE_LS_KEY = 'l10n.bundle';

export async function loadL10nBundle() {
  const bundle = await getLatestL10nBundle().catch(() => ({
    locale: DEFAULT_LOCALE,
    messages: {}
  }));

  resetCurrentLocale(bundle.locale);
  resetMessages(bundle.messages);

  // No need to load english (default) bundle, it's coming with react-intl
  if (bundle.locale !== DEFAULT_LOCALE) {
    const [intlBundle, intl] = await Promise.all([
      import(`react-intl/locale-data/${bundle.locale}`),
      import('react-intl')
    ]);

    intl.addLocaleData(intlBundle.default);
  }

  return bundle;
}

export async function getLatestL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = loadL10nBundleFromLocalStorage();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch(response => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {}
      };
    }
    throw new Error(`Unexpected status code: ${response.status}`);
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages
  };

  saveL10nBundleToLocalStorage(bundle);

  return bundle;
}

export function getCurrentL10nBundle() {
  return loadL10nBundleFromLocalStorage();
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function loadL10nBundleFromLocalStorage() {
  let bundle: L10nBundle;

  try {
    bundle = JSON.parse(loadFromLocalStorage(L10N_BUNDLE_LS_KEY) ?? '{}');
  } catch {
    bundle = {};
  }

  return bundle;
}

function saveL10nBundleToLocalStorage(bundle: L10nBundle) {
  saveInLocalStorage(L10N_BUNDLE_LS_KEY, JSON.stringify(bundle));
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(message);
  }
}
