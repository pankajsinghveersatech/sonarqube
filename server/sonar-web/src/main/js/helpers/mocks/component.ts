/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier, TreeComponent } from '../../types/component';
import { MetricKey } from '../../types/metrics';
import { mockMeasureEnhanced } from '../testMocks';

export function mockComponent(overrides: Partial<T.Component> = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    qualifier: ComponentQualifier.Project,
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockTreeComponent(overrides: Partial<TreeComponent>): TreeComponent {
  return {
    key: 'my-key',
    qualifier: ComponentQualifier.Project,
    name: 'component',
    visibility: 'public',
    ...overrides
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<T.ComponentMeasure> = {}
): T.ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: ComponentQualifier.File,
      path: 'src/index.tsx',
      measures: [{ metric: MetricKey.bugs, value: '1', bestValue: false }],
      ...overrides
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: ComponentQualifier.Project,
    measures: [{ metric: MetricKey.bugs, value: '12', bestValue: false }],
    ...overrides
  };
}

export function mockComponentMeasureEnhanced(
  overrides: Partial<T.ComponentMeasureEnhanced> = {}
): T.ComponentMeasureEnhanced {
  return {
    ...mockComponentMeasure(false, overrides as T.ComponentMeasure),
    leak: undefined,
    measures: [mockMeasureEnhanced()],
    value: undefined,
    ...overrides
  };
}
