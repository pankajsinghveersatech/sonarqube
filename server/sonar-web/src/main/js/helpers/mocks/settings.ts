/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Setting, SettingCategoryDefinition, SettingWithCategory } from '../../types/settings';

export function mockDefinition(
  overrides: Partial<SettingCategoryDefinition> = {}
): SettingCategoryDefinition {
  return {
    key: 'foo',
    category: 'foo category',
    fields: [],
    options: [],
    subCategory: 'foo subCat',
    ...overrides
  };
}

export function mockSetting(overrides: Partial<Setting> = {}): Setting {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: []
    },
    ...overrides
  };
}

export function mockSettingWithCategory(
  overrides: Partial<SettingWithCategory> = {}
): SettingWithCategory {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: [],
      category: 'general',
      fields: [],
      subCategory: 'email'
    },
    ...overrides
  };
}
