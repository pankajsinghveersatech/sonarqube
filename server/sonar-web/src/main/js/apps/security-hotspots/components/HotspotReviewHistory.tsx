/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { Button, ButtonLink, DeleteButton, EditButton } from '../../../components/controls/buttons';
import Dropdown, { DropdownOverlay } from '../../../components/controls/Dropdown';
import Toggler from '../../../components/controls/Toggler';
import DateTimeFormatter from '../../../components/intl/DateTimeFormatter';
import IssueChangelogDiff from '../../../components/issue/components/IssueChangelogDiff';
import Avatar from '../../../components/ui/Avatar';
import { PopupPlacement } from '../../../components/ui/popups';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import { sanitizeString } from '../../../helpers/sanitize';
import { Hotspot, ReviewHistoryType } from '../../../types/security-hotspots';
import { getHotspotReviewHistory } from '../utils';
import HotspotCommentPopup from './HotspotCommentPopup';

export interface HotspotReviewHistoryProps {
  hotspot: Hotspot;
  onDeleteComment: (key: string) => void;
  onEditComment: (key: string, comment: string) => void;
  onShowFullHistory: () => void;
  showFullHistory: boolean;
}

export const MAX_RECENT_ACTIVITY = 5;

export default function HotspotReviewHistory(props: HotspotReviewHistoryProps) {
  const { hotspot, showFullHistory } = props;
  const fullReviewHistory = getHotspotReviewHistory(hotspot);
  const [editedCommentKey, setEditedCommentKey] = React.useState('');

  const reviewHistory = showFullHistory
    ? fullReviewHistory
    : fullReviewHistory.slice(0, MAX_RECENT_ACTIVITY);

  return (
    <>
      <ul>
        {reviewHistory.map((historyElt, historyIndex) => {
          const { user, type, diffs, date, html, key, updatable, markdown } = historyElt;
          return (
            <li
              className={classNames('padded-top padded-bottom', {
                'bordered-top': historyIndex > 0
              })}
              key={historyIndex}>
              <div className="display-flex-center">
                {user.name && (
                  <>
                    <Avatar
                      className="little-spacer-right"
                      hash={user.avatar}
                      name={user.name}
                      size={20}
                    />
                    <strong>
                      {user.active
                        ? user.name
                        : translateWithParameters('user.x_deleted', user.name)}
                    </strong>
                    {type === ReviewHistoryType.Creation && (
                      <span className="little-spacer-left">
                        {translate('hotspots.review_history.created')}
                      </span>
                    )}
                    {type === ReviewHistoryType.Comment && (
                      <span className="little-spacer-left">
                        {translate('hotspots.review_history.comment_added')}
                      </span>
                    )}
                    <span className="little-spacer-left little-spacer-right">-</span>
                  </>
                )}
                <DateTimeFormatter date={date} />
              </div>

              {type === ReviewHistoryType.Diff && diffs && (
                <div className="spacer-top">
                  {diffs.map((diff, diffIndex) => (
                    <IssueChangelogDiff diff={diff} key={diffIndex} />
                  ))}
                </div>
              )}

              {type === ReviewHistoryType.Comment && key && html && markdown && (
                <div className="spacer-top display-flex-space-between">
                  <div
                    className="markdown"
                    // eslint-disable-next-line react/no-danger
                    dangerouslySetInnerHTML={{ __html: sanitizeString(html) }}
                  />
                  {updatable && (
                    <div>
                      <div className="dropdown">
                        <Toggler
                          onRequestClose={() => {
                            setEditedCommentKey('');
                          }}
                          open={key === editedCommentKey}
                          overlay={
                            <DropdownOverlay placement={PopupPlacement.BottomRight}>
                              <HotspotCommentPopup
                                markdownComment={markdown}
                                onCancelEdit={() => setEditedCommentKey('')}
                                onCommentEditSubmit={comment => {
                                  setEditedCommentKey('');
                                  props.onEditComment(key, comment);
                                }}
                              />
                            </DropdownOverlay>
                          }>
                          <EditButton
                            className="button-small"
                            onClick={() => setEditedCommentKey(key)}
                          />
                        </Toggler>
                      </div>
                      <Dropdown
                        onOpen={() => setEditedCommentKey('')}
                        overlay={
                          <div className="padded abs-width-150">
                            <p>{translate('issue.comment.delete_confirm_message')}</p>
                            <Button
                              className="button-red big-spacer-top pull-right"
                              onClick={() => props.onDeleteComment(key)}>
                              {translate('delete')}
                            </Button>
                          </div>
                        }
                        overlayPlacement={PopupPlacement.BottomRight}>
                        <DeleteButton className="button-small" />
                      </Dropdown>
                    </div>
                  )}
                </div>
              )}
            </li>
          );
        })}
      </ul>
      {!showFullHistory && fullReviewHistory.length > MAX_RECENT_ACTIVITY && (
        <ButtonLink className="spacer-top" onClick={props.onShowFullHistory}>
          {translate('show_all')}
        </ButtonLink>
      )}
    </>
  );
}
