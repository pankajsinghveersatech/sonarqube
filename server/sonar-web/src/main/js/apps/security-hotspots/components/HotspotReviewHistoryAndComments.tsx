/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import {
  commentSecurityHotspot,
  deleteSecurityHotspotComment,
  editSecurityHotspotComment
} from '../../../api/security-hotspots';
import FormattingTips from '../../../components/common/FormattingTips';
import { Button } from '../../../components/controls/buttons';
import { translate } from '../../../helpers/l10n';
import { isLoggedIn } from '../../../helpers/users';
import { Hotspot } from '../../../types/security-hotspots';
import HotspotReviewHistory from './HotspotReviewHistory';

interface Props {
  currentUser: T.CurrentUser;
  hotspot: Hotspot;
  commentTextRef: React.RefObject<HTMLTextAreaElement>;
  onCommentUpdate: () => void;
}

interface State {
  comment: string;
  showFullHistory: boolean;
}

export default class HotspotReviewHistoryAndComments extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      comment: '',
      showFullHistory: false
    };
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.hotspot.key !== this.props.hotspot.key) {
      this.setState({
        comment: '',
        showFullHistory: false
      });
    }
  }

  handleCommentChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
    this.setState({ comment: event.target.value });
  };

  handleSubmitComment = () => {
    return commentSecurityHotspot(this.props.hotspot.key, this.state.comment).then(() => {
      this.setState({ comment: '' });
      this.props.onCommentUpdate();
    });
  };

  handleDeleteComment = (key: string) => {
    return deleteSecurityHotspotComment(key).then(() => {
      this.props.onCommentUpdate();
    });
  };

  handleEditComment = (key: string, comment: string) => {
    return editSecurityHotspotComment(key, comment).then(() => {
      this.props.onCommentUpdate();
    });
  };

  handleShowFullHistory = () => {
    this.setState({ showFullHistory: true });
  };

  render() {
    const { currentUser, hotspot, commentTextRef } = this.props;
    const { comment, showFullHistory } = this.state;
    return (
      <div className="padded it__hs-review-history">
        {isLoggedIn(currentUser) && (
          <div className="big-spacer-top">
            <div className="little-spacer-bottom">{translate('hotspots.comment.field')}</div>
            <textarea
              className="form-field fixed-width width-100 spacer-bottom"
              onChange={this.handleCommentChange}
              ref={commentTextRef}
              rows={2}
              value={comment}
            />
            <div className="display-flex-space-between display-flex-center ">
              <FormattingTips className="huge-spacer-bottom" />
              <div>
                <Button
                  className="huge-spacer-bottom"
                  id="hotspot-comment-box-submit"
                  onClick={this.handleSubmitComment}>
                  {translate('hotspots.comment.submit')}
                </Button>
              </div>
            </div>
          </div>
        )}

        <h2 className="spacer-top big-spacer-bottom">{translate('hotspot.section.activity')}</h2>

        <HotspotReviewHistory
          hotspot={hotspot}
          onDeleteComment={this.handleDeleteComment}
          onEditComment={this.handleEditComment}
          onShowFullHistory={this.handleShowFullHistory}
          showFullHistory={showFullHistory}
        />
      </div>
    );
  }
}
