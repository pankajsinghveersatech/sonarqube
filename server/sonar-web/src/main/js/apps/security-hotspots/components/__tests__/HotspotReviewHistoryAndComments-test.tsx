/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import {
  commentSecurityHotspot,
  deleteSecurityHotspotComment,
  editSecurityHotspotComment
} from '../../../../api/security-hotspots';
import { mockHotspot } from '../../../../helpers/mocks/security-hotspots';
import { mockCurrentUser } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import { isLoggedIn } from '../../../../helpers/users';
import HotspotReviewHistory from '../HotspotReviewHistory';
import HotspotReviewHistoryAndComments from '../HotspotReviewHistoryAndComments';

jest.mock('../../../../api/security-hotspots', () => ({
  commentSecurityHotspot: jest.fn().mockResolvedValue({}),
  deleteSecurityHotspotComment: jest.fn().mockResolvedValue({}),
  editSecurityHotspotComment: jest.fn().mockResolvedValue({})
}));

jest.mock('../../../../helpers/users', () => ({ isLoggedIn: jest.fn(() => true) }));

it('should render correctly', () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
});

it('should render correctly without user', () => {
  ((isLoggedIn as any) as jest.Mock<boolean, [boolean]>).mockReturnValueOnce(false);
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();
});

it('should submit comment', async () => {
  const mockApi = commentSecurityHotspot as jest.Mock;
  const hotspot = mockHotspot();
  const wrapper = shallowRender({ hotspot });
  mockApi.mockClear();
  wrapper.instance().setState({ comment: 'Comment' });

  wrapper.find('#hotspot-comment-box-submit').simulate('click');
  await waitAndUpdate(wrapper);

  expect(mockApi).toHaveBeenCalledWith(hotspot.key, 'Comment');
  expect(wrapper.state().comment).toBe('');
  expect(wrapper.instance().props.onCommentUpdate).toHaveBeenCalledTimes(1);
});

it('should change comment', () => {
  const wrapper = shallowRender();
  wrapper.instance().setState({ comment: 'Comment' });
  wrapper.find('textarea').simulate('change', { target: { value: 'Foo' } });

  expect(wrapper.state().comment).toBe('Foo');
});

it('should reset on change hotspot', () => {
  const wrapper = shallowRender();
  wrapper.setState({ comment: 'NOP' });
  wrapper.setProps({ hotspot: mockHotspot({ key: 'other-hotspot' }) });

  expect(wrapper.state().comment).toBe('');
});

it('should delete comment', async () => {
  const wrapper = shallowRender();

  wrapper.find(HotspotReviewHistory).simulate('deleteComment', 'me1');
  await waitAndUpdate(wrapper);

  expect(deleteSecurityHotspotComment).toHaveBeenCalledWith('me1');
  expect(wrapper.instance().props.onCommentUpdate).toBeCalledTimes(1);
});

it('should edit comment', async () => {
  const wrapper = shallowRender();

  wrapper.find(HotspotReviewHistory).simulate('editComment', 'me1', 'new');
  await waitAndUpdate(wrapper);

  expect(editSecurityHotspotComment).toHaveBeenCalledWith('me1', 'new');
  expect(wrapper.instance().props.onCommentUpdate).toBeCalledTimes(1);
});

it('should correctly toggle the show full history state', () => {
  const wrapper = shallowRender();
  expect(wrapper.state().showFullHistory).toBe(false);
  wrapper
    .find(HotspotReviewHistory)
    .props()
    .onShowFullHistory();
  expect(wrapper.state().showFullHistory).toBe(true);
});

function shallowRender(props?: Partial<HotspotReviewHistoryAndComments['props']>) {
  return shallow<HotspotReviewHistoryAndComments>(
    <HotspotReviewHistoryAndComments
      commentTextRef={React.createRef()}
      currentUser={mockCurrentUser()}
      hotspot={mockHotspot()}
      onCommentUpdate={jest.fn()}
      {...props}
    />
  );
}
