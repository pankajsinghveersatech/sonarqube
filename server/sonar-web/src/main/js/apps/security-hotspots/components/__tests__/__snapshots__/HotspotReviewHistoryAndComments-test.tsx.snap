// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="padded it__hs-review-history"
>
  <div
    className="big-spacer-top"
  >
    <div
      className="little-spacer-bottom"
    >
      hotspots.comment.field
    </div>
    <textarea
      className="form-field fixed-width width-100 spacer-bottom"
      onChange={[Function]}
      rows={2}
      value=""
    />
    <div
      className="display-flex-space-between display-flex-center "
    >
      <FormattingTips
        className="huge-spacer-bottom"
      />
      <div>
        <Button
          className="huge-spacer-bottom"
          id="hotspot-comment-box-submit"
          onClick={[Function]}
        >
          hotspots.comment.submit
        </Button>
      </div>
    </div>
  </div>
  <h2
    className="spacer-top big-spacer-bottom"
  >
    hotspot.section.activity
  </h2>
  <HotspotReviewHistory
    hotspot={
      Object {
        "assignee": "assignee",
        "assigneeUser": Object {
          "active": true,
          "local": true,
          "login": "assignee",
          "name": "John Doe",
        },
        "author": "author",
        "authorUser": Object {
          "active": true,
          "local": true,
          "login": "author",
          "name": "John Doe",
        },
        "canChangeStatus": true,
        "changelog": Array [],
        "comment": Array [],
        "component": Object {
          "key": "hotspot-component",
          "longName": "Hotspot component long name",
          "name": "Hotspot Component",
          "path": "path/to/component",
          "qualifier": "FIL",
        },
        "creationDate": "2013-05-13T17:55:41+0200",
        "key": "01fc972e-2a3c-433e-bcae-0bd7f88f5123",
        "line": 142,
        "message": "'3' is a magic number.",
        "project": Object {
          "key": "hotspot-component",
          "longName": "Hotspot component long name",
          "name": "Hotspot Component",
          "path": "path/to/component",
          "qualifier": "TRK",
        },
        "resolution": "FIXED",
        "rule": Object {
          "fixRecommendations": "<p>This a <strong>strong</strong> message about fixing !</p>",
          "key": "squid:S2077",
          "name": "That rule",
          "riskDescription": "<p>This a <strong>strong</strong> message about risk !</p>",
          "securityCategory": "sql-injection",
          "vulnerabilityDescription": "<p>This a <strong>strong</strong> message about vulnerability !</p>",
          "vulnerabilityProbability": "HIGH",
        },
        "status": "REVIEWED",
        "textRange": Object {
          "endLine": 142,
          "endOffset": 83,
          "startLine": 142,
          "startOffset": 26,
        },
        "updateDate": "2013-05-13T17:55:42+0200",
        "users": Array [
          Object {
            "active": true,
            "local": true,
            "login": "assignee",
            "name": "John Doe",
          },
          Object {
            "active": true,
            "local": true,
            "login": "author",
            "name": "John Doe",
          },
        ],
      }
    }
    onDeleteComment={[Function]}
    onEditComment={[Function]}
    onShowFullHistory={[Function]}
    showFullHistory={false}
  />
</div>
`;

exports[`should render correctly without user 1`] = `
<div
  className="padded it__hs-review-history"
>
  <h2
    className="spacer-top big-spacer-bottom"
  >
    hotspot.section.activity
  </h2>
  <HotspotReviewHistory
    hotspot={
      Object {
        "assignee": "assignee",
        "assigneeUser": Object {
          "active": true,
          "local": true,
          "login": "assignee",
          "name": "John Doe",
        },
        "author": "author",
        "authorUser": Object {
          "active": true,
          "local": true,
          "login": "author",
          "name": "John Doe",
        },
        "canChangeStatus": true,
        "changelog": Array [],
        "comment": Array [],
        "component": Object {
          "key": "hotspot-component",
          "longName": "Hotspot component long name",
          "name": "Hotspot Component",
          "path": "path/to/component",
          "qualifier": "FIL",
        },
        "creationDate": "2013-05-13T17:55:41+0200",
        "key": "01fc972e-2a3c-433e-bcae-0bd7f88f5123",
        "line": 142,
        "message": "'3' is a magic number.",
        "project": Object {
          "key": "hotspot-component",
          "longName": "Hotspot component long name",
          "name": "Hotspot Component",
          "path": "path/to/component",
          "qualifier": "TRK",
        },
        "resolution": "FIXED",
        "rule": Object {
          "fixRecommendations": "<p>This a <strong>strong</strong> message about fixing !</p>",
          "key": "squid:S2077",
          "name": "That rule",
          "riskDescription": "<p>This a <strong>strong</strong> message about risk !</p>",
          "securityCategory": "sql-injection",
          "vulnerabilityDescription": "<p>This a <strong>strong</strong> message about vulnerability !</p>",
          "vulnerabilityProbability": "HIGH",
        },
        "status": "REVIEWED",
        "textRange": Object {
          "endLine": 142,
          "endOffset": 83,
          "startLine": 142,
          "startOffset": 26,
        },
        "updateDate": "2013-05-13T17:55:42+0200",
        "users": Array [
          Object {
            "active": true,
            "local": true,
            "login": "assignee",
            "name": "John Doe",
          },
          Object {
            "active": true,
            "local": true,
            "login": "author",
            "name": "John Doe",
          },
        ],
      }
    }
    onDeleteComment={[Function]}
    onEditComment={[Function]}
    onShowFullHistory={[Function]}
    showFullHistory={false}
  />
</div>
`;
